qiskit_experiments/analysis/curve_fitting.py

```diff
@@ -0,0 +1,220 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Curve fitting functions for experiment analysis
+"""
+# pylint: disable = invalid-name
+
+from typing import List, Dict, Tuple, Callable, Optional
+
+import numpy as np
+import scipy.optimize as opt
+
+from qiskit.exceptions import QiskitError
+from qiskit_experiments.base_analysis import AnalysisResult
+from qiskit_experiments.analysis.data_processing import filter_data
+
+
+def curve_fit(
+    func: Callable,
+    xdata: np.ndarray,
+    ydata: np.ndarray,
+    p0: np.ndarray,
+    sigma: Optional[np.ndarray] = None,
+    **kwargs,
+) -> AnalysisResult:
+    """Use non-linear least squares to fit a function, f, to data
+
+    Wraps scipy.optimize.curve_fit
+
+    Args:
+        func: a fit function `f(x *params)`.
+        xdata: a 1D array of x-data
+        ydata: a 1D array of y-data
+        p0: initial guess for optimization parameters.
+        sigma: Optional, a 1D array of standard deviations in ydata.
+        kwargs: additional kwargs for scipy.optimize.curve_fit.
+
+    Returns:
+        AnalysisResult: result containing `popt` the optimal fit parameters,
+                        `popt_err` the standard error estimates popt,
+                        `pcov` the covariance matrix for the fit,
+                        `chisq` the chi-squared parameter of fit,
+                        `xrange` the range of xdata values used for fit.
+    """
+
+    # Run curve fit
+    # pylint: disable = unbalanced-tuple-unpacking
+    popt, pcov = opt.curve_fit(func, xdata, ydata, sigma=sigma, p0=p0, **kwargs)
+    popt_err = np.sqrt(np.diag(pcov))
+
+    # Compute chi-squared for fit
+    yfits = func(xdata, *popt)
+    chisq = np.mean(((yfits - ydata) / sigma) ** 2)
+
+    # Compute xdata range for fit
+    xdata_range = [min(xdata), max(xdata)]
+
+    result = {
+        "popt": popt,
+        "popt_err": popt_err,
+        "pcov": pcov,
+        "chisq": chisq,
+        "xrange": xdata_range,
+    }
+    # TODO:
+    #  1. Add some basic validation of computer good / bad based on fit result.
+    #  2. Add error handling so if fitting fails we can return an analysis
+    #     result containing this information
+    return AnalysisResult(result)
+
+
+def multi_curve_fit(
+    funcs: List[Callable],
+    xdata: np.ndarray,
+    ydata: np.ndarray,
+    p0: np.ndarray,
+    sigma: Optional[np.ndarray] = None,
+    weights: Optional[np.ndarray] = None,
+    **kwargs,
+):
+    """Use non-linear least squares to fit a list of functions, f_i, to data
```

Comment:
Can you make this more precise? It is not clear to me what is happening here. Are we fitting each row of the `xdata` to its own fit function?